packages/@aws-cdk/aws-iam/lib/immutable-role.ts

```diff
@@ -0,0 +1,43 @@
+import { Grant } from './grant';
+import { IManagedPolicy } from './managed-policy';
+import { Policy, PolicyProps } from './policy';
+import { PolicyStatement } from './policy-statement';
+import { IPrincipal } from './principals';
+import { IRole } from './role';
+
+export class ImmutableRole implements IRole {
+  public readonly assumeRoleAction = this.role.assumeRoleAction;
+  public readonly policyFragment = this.role.policyFragment;
+  public readonly grantPrincipal = this.role.grantPrincipal;
+  public readonly roleArn = this.role.roleArn;
+  public readonly roleName = this.role.roleName;
+  public readonly node = this.role.node;
+  public readonly stack = this.role.stack;
+
+  constructor(private readonly role: IRole) {
+  }
+
+  public createAndAttachPolicy(_id: string, _props?: PolicyProps): Policy | undefined {
+    return undefined;
+  }
+
+  public attachInlinePolicy(_policy: Policy): void {
+    // do nothing
+  }
+
+  public addManagedPolicy(_policy: IManagedPolicy): void {
+    // do nothing
+  }
+
+  public addToPolicy(_statement: PolicyStatement): boolean {
+    return false;
```

Comment:
I think this might need to return `true`. I'm sorry :(.

Reply:
I don't understand this comment. Can you elaborate?

Reply:
```ts
  /**
   * Add to the policy of this principal.
   *
   * @returns true if the statement was added, false if the principal in
   * question does not have a policy document to add the statement to.
   */
  addToPolicy(statement: PolicyStatement): boolean;
```

I know that the statement wasn't *really* added, but `false` means that a policy statement could NEVER be added and one should be added to the resource instead.

In this case I guess it's a toss-up between "the user has already prepared all statements so don't do anything" and "this role is immutable so add to the resource instead."

I would say we either make it configurable between those 2, or default to the do-less option (with an option to make it configurable later). Right now, the do-less default is returning "true" (pretend everything worked fine attaching to the principal).

Reply:
Shouldn't this be `true` here as well then...? (for roles imported with `{ mutable: false }`) https://github.com/aws/aws-cdk/blob/ddb05f95a372d563abfdaed34e1acecb0772133b/packages/@aws-cdk/aws-iam/lib/role.ts#L220-L228

Reply:
Yes it should.